### PR TITLE
Sensorthings API Provider

### PR DESCRIPTION
**Pull request by webb-ben**
This is a PR based off of #685, adding a SensorThing API (STA) provider to the core pygeoapi feature providers. The provider is capable of creating feature collections for the STA entities: Thing (with Location), Datastream (with ObservedProperty and Sensor), and Observation. Thus with a pygeoapi configuration with a feature collection for each of these entities, all STA entities of a given STA endpoint can be represented in pygeoapi. In this way, pending STA/OAFeature alignment, STA endpoints can be proxied to conform to OGC API - Features via pygeoapi.

The STA provider introduces two new fields in the provider configuration: ``entity`` and ``intralink``. The ``entity`` field is required and corresponds to which entity of Thing, Datastream, or Observation is used for the feature collection. The ``intralink`` field is a boolean value specifying whether items in that feature collection should be linked to by the other feature collections or not. If ``intralink`` is true for Thing but not Datastream: A Thing item will include the attributes of its associated Datastreams, while a Datastream item will replace its related Thing attributes with a link to the pygeoapi URI of the linked Thing item. The provider adds ``sosa`` and ``ssn`` context items sufficient to express linkages between Things (as Features of Interest), Datastreams (as ObservationCollections) and Observations (as Observations) in the JSON-LD output format.The PR also includes documentation for the STA provider, docker examples, and pytests for all of the query parameters.

### CODE REVIEW

**Comment by ksonda** [comment]
@tomkralidis This is ready for review although I don't know who is best to, maybe @justb4 ?

**Comment by tomkralidis** [comment]
@ksonda thanks.  I've put @justb4 and myself for now, and happy for others to review who have time.

**Comment by glennlaughlin** [comment]
> @ksonda thanks. I've put @justb4 and myself for now, and happy for others to review who have time.

Hi Tom (et al). I'd be happy to review this if appropriate.  I've been reviewing the OGC O&M for (arguably too long) a while now and I'm interested to see where SensorThings aligns generally. I haven't been involved with pygeoapi so not sure my comments would be too inciteful but happy to spend a few cycles on this if it doesnt' distract from the project.

**Comment by ksonda** [comment]
@glennlaughlin There was more of a discussion about what to do with forcing the STA data model into the OAF pattern in #685. This PR just implements the pattern we settled on. Feel free to look and contribute over there too!

**Comment by glennlaughlin** [comment]
@ksonda OK. Sounds like a good place to start. Thx.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -75,6 +80,7 @@ jobs:
         pytest tests/test_openapi.py
         #pytest tests/test_postgresql_provider.py
         pytest tests/test_rasterio_provider.py
+        pytest tests/test_sensorthings__provider.py
```
Comment (tomkralidis) [inline .github/workflows/main.yml:83]: Rename file to `tests/test_sensorthings_provider.py `.

File: docker/examples/sensorthings/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,3 @@
+FROM webbben/pygeoapi-sensorthings:latest
```
Comment (tomkralidis) [inline docker/examples/sensorthings/Dockerfile:1]: Can you add a `README.md` in `docker/examples/sensorthings` with some high level notes (and pointer to the codebase/setup of `webbben/pygeoapi-sensorthings` ?

File: pygeoapi/plugin.py [hunk 3]
```diff
@@ -50,6 +50,7 @@
         'xarray': 'pygeoapi.provider.xarray_.XarrayProvider',
         'MVT': 'pygeoapi.provider.mvt.MVTProvider',
         'TinyDBCatalogue': 'pygeoapi.provider.tinydb_.TinyDBCatalogueProvider',
+        'Sensorthings': 'pygeoapi.provider.sensorthings.SensorthingsProvider',
```
Comment (tomkralidis) [inline pygeoapi/plugin.py:53]: (global comment) should this be `SensorThings` (vs. `Sensorthings`)?

**Review by tomkralidis — CHANGES_REQUESTED** [review]
Minor changes requested -- great work!

File: docker/examples/sensorthings/Dockerfile (hunk 2 above)
Comment (webb-ben) [reply docker/examples/sensorthings/Dockerfile:1]: `webbben/pygeoapi-sensorthings` points to my branch of pygeoapi with the Sensorthings provider. With a merged PR this is redundant. That being said, I will add a readme and think I am going to replace the Dockerfile with a docker-compose file.

File: pygeoapi/plugin.py (hunk 3 above)
Comment (webb-ben) [reply pygeoapi/plugin.py:53]: Probably `SensorThings`

**Comment by webb-ben** [comment]
@tomkralidis I believe I have made all of the adjustments you requested. The only issue I am running into now, is that the github workflow builds are failing for all branches because of something in test_collection_coverage. Here is the [link](https://github.com/webb-ben/pygeoapi/actions), if you have any suggestions.

**Comment by tomkralidis** [comment]
Looks like this is an unrelated CI issue.

**Comment by tomkralidis** [comment]
Great work!!

**Pull request merged**